File: opennow-stable/src/renderer/src/gfn/sdp.ts
import type { VideoCodec } from "@shared/gfn";

interface IceCredentials {
  ufrag: string;
  pwd: string;
  fingerprint: string;
}

/**
 * Convert dash-separated hostname to dotted IP if it matches the GFN pattern.
 * e.g. "80-250-97-40.cloudmatchbeta.nvidiagrid.net" -> "80.250.97.40"
 * e.g. "161-248-11-132.bpc.geforcenow.nvidiagrid.net" -> "161.248.11.132"
 */
export function extractPublicIp(hostOrIp: string): string | null {
  if (!hostOrIp) return null;

  // Already a dotted IP?
  if (/^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/.test(hostOrIp)) {
    return hostOrIp;
  }

  // Dash-separated hostname: take the first label, convert dashes to dots
  const firstLabel = hostOrIp.split(".")[0] ?? "";
  const parts = firstLabel.split("-");
  if (parts.length === 4 && parts.every((p) => /^\d{1,3}$/.test(p))) {
    return parts.join(".");
  }

  return null;
}

/**
 * Fix 0.0.0.0 in the server's SDP offer with the actual server IP.
 * Matches Rust's fix_server_ip() — replaces "c=IN IP4 0.0.0.0" with real IP.
 * Also fixes a=candidate: lines that contain 0.0.0.0 as the candidate IP,
 * since Chrome's WebRTC stack treats those as unreachable and ICE fails.
 */
export function fixServerIp(sdp: string, serverIp: string): string {
  const ip = extractPublicIp(serverIp);
  if (!ip) {
    console.log(`[SDP] fixServerIp: could not extract IP from "${serverIp}"`);
    return sdp;
  }
  // 1. Fix connection lines: c=IN IP4 0.0.0.0
  const cCount = (sdp.match(/c=IN IP4 0\.0\.0\.0/g) ?? []).length;
  let fixed = sdp.replace(/c=IN IP4 0\.0\.0\.0/g, `c=IN IP4 ${ip}`);
  console.log(`[SDP] fixServerIp: replaced ${cCount} c= lines with ${ip}`);

  // 2. Fix ICE candidate lines: a=candidate:... 0.0.0.0 ...
  //    Format: a=candidate:<foundation> <component> <protocol> <priority> <ip> <port> typ <type>
  const candidateCount = (fixed.match(/(a=candidate:\S+\s+\d+\s+\w+\s+\d+\s+)0\.0\.0\.0(\s+)/g) ?? []).length;
  if (candidateCount > 0) {
    fixed = fixed.replace(
      /(a=candidate:\S+\s+\d+\s+\w+\s+\d+\s+)0\.0\.0\.0(\s+)/g,
      `$1${ip}$2`,
    );
    console.log(`[SDP] fixServerIp: replaced ${candidateCount} a=candidate lines with ${ip}`);
  }

  return fixed;
}

/**
 * Extract the server's ice-ufrag from the offer SDP.
 * Needed for manual ICE candidate injection (ice-lite servers).
 */
export function extractIceUfragFromOffer(sdp: string): string {
  const match = sdp.match(/a=ice-ufrag:([^\r\n]+)/);
  return match?.[1]?.trim() ?? "";
}

export function extractIceCredentials(sdp: string): IceCredentials {
  const ufrag = sdp
    .split(/\r?\n/)
    .find((line) => line.startsWith("a=ice-ufrag:"))
    ?.replace("a=ice-ufrag:", "")
    .trim();
  const pwd = sdp
    .split(/\r?\n/)
    .find((line) => line.startsWith("a=ice-pwd:"))
    ?.replace("a=ice-pwd:", "")
    .trim();
  const fingerprint = sdp
    .split(/\r?\n/)
    .find((line) => line.startsWith("a=fingerprint:sha-256 "))
    ?.replace("a=fingerprint:sha-256 ", "")
    .trim();

  return {
    ufrag: ufrag ?? "",
    pwd: pwd ?? "",
    fingerprint: fingerprint ?? "",
  };
}

function normalizeCodec(name: string): string {
  const upper = name.toUpperCase();
  return upper === "HEVC" ? "H265" : upper;
}

export function preferCodec(sdp: string, codec: VideoCodec): string {
  console.log(`[SDP] preferCodec: filtering SDP for codec "${codec}"`);
  const lineEnding = sdp.includes("\r\n") ? "\r\n" : "\n";
  const lines = sdp.split(/\r?\n/);

  let inVideoSection = false;
  const payloadTypesByCodec = new Map<string, string[]>();

  for (const line of lines) {
    if (line.startsWith("m=video")) {
      inVideoSection = true;
      continue;
    }
    if (line.startsWith("m=") && inVideoSection) {
      inVideoSection = false;
    }

    if (!inVideoSection || !line.startsWith("a=rtpmap:")) {
      continue;
    }

    const [, rest = ""] = line.split("a=rtpmap:");
    const [pt, codecPart] = rest.split(/\s+/, 2);
    const codecName = normalizeCodec((codecPart ?? "").split("/")[0] ?? "");
    if (!pt || !codecName) {
      continue;
    }

    const list = payloadTypesByCodec.get(codecName) ?? [];
    list.push(pt);
    payloadTypesByCodec.set(codecName, list);
  }

  // Log all codecs found in the SDP
  for (const [name, pts] of payloadTypesByCodec.entries()) {
    console.log(`[SDP] preferCodec: found codec ${name} with payload types [${pts.join(", ")}]`);
  }

  const preferred = new Set(payloadTypesByCodec.get(codec) ?? []);
  if (preferred.size === 0) {
    console.log(`[SDP] preferCodec: codec "${codec}" NOT found in offer — returning SDP unmodified`);
    return sdp;
  }

  console.log(`[SDP] preferCodec: keeping payload types [${Array.from(preferred).join(", ")}] for ${codec}`);

  const filtered: string[] = [];
  inVideoSection = false;

  for (const line of lines) {
    if (line.startsWith("m=video")) {
      inVideoSection = true;
      const parts = line.split(/\s+/);
      const header = parts.slice(0, 3);
      const payloads = parts.slice(3).filter((pt) => preferred.has(pt));
      filtered.push(payloads.length > 0 ? [...header, ...payloads].join(" ") : line);
      continue;
    }

    if (line.startsWith("m=") && inVideoSection) {
      inVideoSection = false;
    }

    if (inVideoSection) {
      if (
        line.startsWith("a=rtpmap:") ||
        line.startsWith("a=fmtp:") ||
        line.startsWith("a=rtcp-fb:")
      ) {
        const [, rest = ""] = line.split(":", 2);
        const [pt = ""] = rest.split(/\s+/, 1);
        if (pt && !preferred.has(pt)) {
          continue;
        }
      }
    }

    filtered.push(line);
  }

  return filtered.join(lineEnding);
}

interface NvstParams {
  width: number;
  height: number;
  fps: number;
  maxBitrateKbps: number;
  credentials: IceCredentials;
}

export function buildNvstSdp(params: NvstParams): string {
  console.log(`[SDP] buildNvstSdp: ${params.width}x${params.height}@${params.fps}fps, maxBitrate=${params.maxBitrateKbps}kbps`);
  console.log(`[SDP] buildNvstSdp: ICE ufrag=${params.credentials.ufrag}, pwd=${params.credentials.pwd.slice(0, 8)}..., fingerprint=${params.credentials.fingerprint.slice(0, 20)}...`);
  const minBitrate = Math.min(10000, Math.floor(params.maxBitrateKbps / 10));
  const isHighFps = params.fps >= 90;
  const is120Fps = params.fps === 120;
  const is240Fps = params.fps >= 240;

  const lines: string[] = [
    "v=0",
    "o=SdpTest test_id_13 14 IN IPv4 127.0.0.1",
    "s=-",
    "t=0 0",
    `a=general.icePassword:${params.credentials.pwd}`,
    `a=general.iceUserNameFragment:${params.credentials.ufrag}`,
    `a=general.dtlsFingerprint:${params.credentials.fingerprint}`,
    "m=video 0 RTP/AVP",
    "a=msid:fbc-video-0",
    // FEC settings
    "a=vqos.fec.rateDropWindow:10",
    "a=vqos.fec.minRequiredFecPackets:2",
    "a=vqos.fec.repairMinPercent:5",
    "a=vqos.fec.repairPercent:5",
    "a=vqos.fec.repairMaxPercent:35",
    // DRC — always disabled to allow full bitrate
    "a=vqos.drc.enable:0",
  ];

  // DFC (Dynamic Frame Control) for high FPS
  if (isHighFps) {
    lines.push(
      "a=vqos.dfc.enable:1",
      "a=vqos.dfc.decodeFpsAdjPercent:85",
      "a=vqos.dfc.targetDownCooldownMs:250",
      "a=vqos.dfc.dfcAlgoVersion:2",
      `a=vqos.dfc.minTargetFps:${is120Fps ? 100 : 60}`,
    );
  }

  // Video encoder settings
  lines.push(
    "a=video.dx9EnableNv12:1",
    "a=video.dx9EnableHdr:1",
    "a=vqos.qpg.enable:1",
    "a=vqos.resControl.qp.qpg.featureSetting:7",
    "a=bwe.useOwdCongestionControl:1",
    "a=video.enableRtpNack:1",
    "a=vqos.bw.txRxLag.minFeedbackTxDeltaMs:200",
    "a=vqos.drc.bitrateIirFilterFactor:18",
    "a=video.packetSize:1140",
    "a=packetPacing.minNumPacketsPerGroup:15",
  );

  // High FPS optimizations
  if (isHighFps) {
    lines.push(
      "a=bwe.iirFilterFactor:8",
      "a=video.encoderFeatureSetting:47",
      "a=video.encoderPreset:6",
      "a=vqos.resControl.cpmRtc.badNwSkipFramesCount:600",
      "a=vqos.resControl.cpmRtc.decodeTimeThresholdMs:9",
      `a=video.fbcDynamicFpsGrabTimeoutMs:${is120Fps ? 6 : 18}`,
      `a=vqos.resControl.cpmRtc.serverResolutionUpdateCoolDownCount:${is120Fps ? 6000 : 12000}`,
    );
  }

  // 240+ FPS optimizations
  if (is240Fps) {
    lines.push(
      "a=video.enableNextCaptureMode:1",
      "a=vqos.maxStreamFpsEstimate:240",
      "a=video.videoSplitEncodeStripsPerFrame:3",
      "a=video.updateSplitEncodeStateDynamically:1",
    );
  }

  // Out-of-focus handling + disable ALL dynamic resolution control
  lines.push(
    "a=vqos.adjustStreamingFpsDuringOutOfFocus:1",
    "a=vqos.resControl.cpmRtc.ignoreOutOfFocusWindowState:1",
    "a=vqos.resControl.perfHistory.rtcIgnoreOutOfFocusWindowState:1",
    // Disable CPM-based resolution changes (prevents SSRC switches)
    "a=vqos.resControl.cpmRtc.featureMask:0",
    "a=vqos.resControl.cpmRtc.enable:0",
    // Never scale down resolution
    "a=vqos.resControl.cpmRtc.minResolutionPercent:100",
    // Infinite cooldown to prevent resolution changes
    "a=vqos.resControl.cpmRtc.resolutionChangeHoldonMs:999999",
    // Packet pacing
    `a=packetPacing.numGroups:${is120Fps ? 3 : 5}`,
    "a=packetPacing.maxDelayUs:1000",
    "a=packetPacing.minNumPacketsFrame:10",
    // NACK queue settings
    "a=video.rtpNackQueueLength:1024",
    "a=video.rtpNackQueueMaxPackets:512",
    "a=video.rtpNackMaxPacketCount:25",
    // Resolution/quality thresholds — high values prevent downscaling
    "a=vqos.drc.qpMaxResThresholdAdj:4",
    "a=vqos.grc.qpMaxResThresholdAdj:4",
    "a=vqos.drc.iirFilterFactor:100",
  );

  // Viewport, FPS, and bitrate
  lines.push(
    `a=video.clientViewportWd:${params.width}`,
    `a=video.clientViewportHt:${params.height}`,
    `a=video.maxFPS:${params.fps}`,
    `a=video.initialBitrateKbps:${Math.floor((params.maxBitrateKbps * 3) / 4)}`,
    `a=video.initialPeakBitrateKbps:${params.maxBitrateKbps}`,
    `a=vqos.bw.maximumBitrateKbps:${params.maxBitrateKbps}`,
    `a=vqos.bw.minimumBitrateKbps:${minBitrate}`,
    `a=vqos.bw.peakBitrateKbps:${params.maxBitrateKbps}`,
    `a=vqos.bw.serverPeakBitrateKbps:${params.maxBitrateKbps}`,
    "a=vqos.bw.enableBandwidthEstimation:1",
    "a=vqos.bw.disableBitrateLimit:1",
    // GRC — disabled
    `a=vqos.grc.maximumBitrateKbps:${params.maxBitrateKbps}`,
    "a=vqos.grc.enable:0",
    // Encoder settings
    "a=video.maxNumReferenceFrames:4",
    "a=video.mapRtpTimestampsToFrames:1",
    "a=video.encoderCscMode:3",
    // Disable server-side scaling and prefilter (prevents resolution downgrade)
    "a=video.scalingFeature1:0",
    "a=video.prefilterParams.prefilterModel:0",
    // Audio track
    "m=audio 0 RTP/AVP",
    "a=msid:audio",
    // Mic track
    "m=mic 0 RTP/AVP",
    "a=msid:mic",
    // Input/application track
    "m=application 0 RTP/AVP",
    "a=msid:input_1",
    "a=ri.partialReliableThresholdMs:300",
    "",
  );

  return lines.join("\n");
}

File: native/core/src/sdp.cpp
#include "opennow/sdp.hpp"

#include <regex>
#include <sstream>
#include <string_view>
#include <vector>

#include "opennow/input_protocol.hpp"
#include "opennow/settings.hpp"

namespace opennow {

namespace {

std::vector<std::string> split_lines(const std::string& value) {
  std::vector<std::string> lines;
  std::string current;
  for (const char ch : value) {
    if (ch == '\r') {
      continue;
    }
    if (ch == '\n') {
      lines.push_back(current);
      current.clear();
    } else {
      current.push_back(ch);
    }
  }
  if (!current.empty() || (!value.empty() && value.back() == '\n')) {
    lines.push_back(current);
  }
  return lines;
}

std::string join_lines(const std::vector<std::string>& lines, const std::string& ending) {
  std::ostringstream out;
  for (std::size_t i = 0; i < lines.size(); ++i) {
    if (i > 0) {
      out << ending;
    }
    out << lines[i];
  }
  return out.str();
}

bool starts_with(std::string_view value, std::string_view prefix) {
  return value.substr(0, prefix.size()) == prefix;
}

std::string line_ending_for(const std::string& sdp) {
  return sdp.find("\r\n") == std::string::npos ? "\n" : "\r\n";
}

std::string replace_all(std::string value, const std::string& from, const std::string& to) {
  std::size_t position = 0;
  while ((position = value.find(from, position)) != std::string::npos) {
    value.replace(position, from.size(), to);
    position += to.size();
  }
  return value;
}

}  // namespace

std::optional<std::string> extract_public_ip(const std::string& host_or_ip) {
  static const std::regex dotted_ip(R"(^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$)");
  if (std::regex_match(host_or_ip, dotted_ip)) {
    return host_or_ip;
  }

  const auto first_dot = host_or_ip.find('.');
  const auto first_label = host_or_ip.substr(0, first_dot);
  static const std::regex dashed_ip(R"(^(\d{1,3})-(\d{1,3})-(\d{1,3})-(\d{1,3})$)");
  std::smatch match;
  if (std::regex_match(first_label, match, dashed_ip)) {
    return match[1].str() + "." + match[2].str() + "." + match[3].str() + "." + match[4].str();
  }

  return std::nullopt;
}

std::string fix_server_ip(const std::string& sdp, const std::string& server_ip) {
  const auto ip = extract_public_ip(server_ip);
  if (!ip) {
    return sdp;
  }

  auto fixed = replace_all(sdp, "c=IN IP4 0.0.0.0", "c=IN IP4 " + *ip);
  fixed = replace_all(fixed, " 0.0.0.0 ", " " + *ip + " ");
  return fixed;
}

std::string extract_ice_ufrag_from_offer(const std::string& sdp) {
  static const std::regex pattern(R"(a=ice-ufrag:([^\r\n]+))");
  std::smatch match;
  return std::regex_search(sdp, match, pattern) ? match[1].str() : "";
}

IceCredentials extract_ice_credentials(const std::string& sdp) {
  IceCredentials credentials;
  for (const auto& line : split_lines(sdp)) {
    if (starts_with(line, "a=ice-ufrag:")) {
      credentials.ufrag = line.substr(12);
    } else if (starts_with(line, "a=ice-pwd:")) {
      credentials.pwd = line.substr(10);
    } else if (starts_with(line, "a=fingerprint:sha-256 ")) {
      credentials.fingerprint = line.substr(std::string("a=fingerprint:sha-256 ").size());
    }
  }
  return credentials;
}

RewriteResult rewrite_h265_tier_flag(const std::string& sdp, int tier_flag) {
  const auto ending = line_ending_for(sdp);
  auto lines = split_lines(sdp);
  std::vector<std::string> h265_payloads;
  bool in_video = false;

  for (const auto& line : lines) {
    if (starts_with(line, "m=video")) {
      in_video = true;
      continue;
    }
    if (starts_with(line, "m=") && in_video) {
      in_video = false;
    }
    if (!in_video || !starts_with(line, "a=rtpmap:")) {
      continue;
    }
    const auto colon = line.find(':');
    const auto space = line.find(' ', colon + 1);
    if (colon == std::string::npos || space == std::string::npos) {
      continue;
    }
    const auto payload = line.substr(colon + 1, space - colon - 1);
    const auto codec_start = space + 1;
    const auto slash = line.find('/', codec_start);
    const auto codec = line.substr(codec_start, slash - codec_start);
    if (codec == "H265" || codec == "HEVC" || codec == "h265" || codec == "hevc") {
      h265_payloads.push_back(payload);
    }
  }

  int replacements = 0;
  for (auto& line : lines) {
    if (!starts_with(line, "a=fmtp:")) {
      continue;
    }
    const auto colon = line.find(':');
    const auto space = line.find(' ', colon + 1);
    if (colon == std::string::npos || space == std::string::npos) {
      continue;
    }
    const auto payload = line.substr(colon + 1, space - colon - 1);
    bool is_h265 = false;
    for (const auto& candidate : h265_payloads) {
      if (candidate == payload) {
        is_h265 = true;
        break;
      }
    }
    if (!is_h265) {
      continue;
    }
    const auto next = std::regex_replace(
        line,
        std::regex("tier-flag=1", std::regex_constants::icase),
        "tier-flag=" + std::to_string(tier_flag));
    if (next != line) {
      ++replacements;
      line = next;
    }
  }

  return {join_lines(lines, ending), replacements};
}

std::string munge_answer_sdp(const std::string& sdp, int max_bitrate_kbps) {
  const auto ending = line_ending_for(sdp);
  const auto lines = split_lines(sdp);
  std::vector<std::string> result;

  for (std::size_t i = 0; i < lines.size(); ++i) {
    auto line = lines[i];
    if (starts_with(line, "a=fmtp:") && line.find("minptime=") != std::string::npos &&
        line.find("stereo=1") == std::string::npos) {
      line += ";stereo=1";
    }
    result.push_back(line);

    if (starts_with(line, "m=video") || starts_with(line, "m=audio")) {
      const auto next_line = i + 1 < lines.size() ? lines[i + 1] : "";
      if (!starts_with(next_line, "b=")) {
        result.push_back(
            "b=AS:" + std::to_string(starts_with(line, "m=video") ? max_bitrate_kbps : 128));
      }
    }
  }

  return join_lines(result, ending);
}

std::string build_nvst_sdp(const NvstSdpParams& params) {
  const auto min_bitrate = std::max(5000, static_cast<int>(params.max_bitrate_kbps * 0.35));
  const auto initial_bitrate = std::max(min_bitrate, static_cast<int>(params.max_bitrate_kbps * 0.7));
  const auto bit_depth = color_quality_is_10_bit(params.color_quality) ? 10 : 8;
  const auto hid_device_mask =
      params.hid_device_mask.value_or(kPartiallyReliableHidDeviceMaskAll);
  const auto pr_gamepad = params.enable_partially_reliable_transfer_gamepad.value_or(
      kPartiallyReliableGamepadMaskAll);
  const auto pr_hid =
      params.enable_partially_reliable_transfer_hid.value_or(hid_device_mask);

  std::vector<std::string> lines = {
      "v=0",
      "o=SdpTest test_id_13 14 IN IPv4 127.0.0.1",
      "s=-",
      "t=0 0",
      "a=general.icePassword:" + params.credentials.pwd,
      "a=general.iceUserNameFragment:" + params.credentials.ufrag,
      "a=general.dtlsFingerprint:" + params.credentials.fingerprint,
      "m=video 0 RTP/AVP",
      "a=msid:fbc-video-0",
      "a=vqos.fec.rateDropWindow:10",
      "a=vqos.fec.minRequiredFecPackets:2",
      "a=vqos.fec.repairMinPercent:5",
      "a=vqos.fec.repairPercent:5",
      "a=vqos.fec.repairMaxPercent:35",
      "a=vqos.drc.enable:0",
      "a=vqos.dfc.enable:0",
      "a=video.dx9EnableNv12:1",
      "a=video.dx9EnableHdr:1",
      "a=vqos.qpg.enable:1",
      "a=bwe.useOwdCongestionControl:1",
      "a=video.enableRtpNack:1",
      "a=video.clientViewportWd:" + std::to_string(params.width),
      "a=video.clientViewportHt:" + std::to_string(params.height),
      "a=video.maxFPS:" + std::to_string(params.fps),
      "a=video.initialBitrateKbps:" + std::to_string(initial_bitrate),
      "a=video.initialPeakBitrateKbps:" + std::to_string(params.max_bitrate_kbps),
      "a=vqos.bw.maximumBitrateKbps:" + std::to_string(params.max_bitrate_kbps),
      "a=vqos.bw.minimumBitrateKbps:" + std::to_string(min_bitrate),
      "a=vqos.bw.peakBitrateKbps:" + std::to_string(params.max_bitrate_kbps),
      "a=vqos.bw.serverPeakBitrateKbps:" + std::to_string(params.max_bitrate_kbps),
      "a=vqos.bw.enableBandwidthEstimation:1",
      "a=vqos.bw.disableBitrateLimit:0",
      "a=vqos.grc.maximumBitrateKbps:" + std::to_string(params.max_bitrate_kbps),
      "a=vqos.grc.enable:0",
      "a=video.bitDepth:" + std::to_string(bit_depth),
      std::string("a=video.scalingFeature1:") + (params.codec == VideoCodec::AV1 ? "1" : "0"),
      "m=audio 0 RTP/AVP",
      "a=msid:audio",
      "m=mic 0 RTP/AVP",
      "a=msid:mic",
      "a=rtpmap:0 PCMU/8000",
      "m=application 0 RTP/AVP",
      "a=msid:input_1",
      "a=ri.partialReliableThresholdMs:" + std::to_string(params.partial_reliable_threshold_ms),
      "a=ri.hidDeviceMask:" + std::to_string(hid_device_mask),
      "a=ri.enablePartiallyReliableTransferGamepad:" + std::to_string(pr_gamepad),
      "a=ri.enablePartiallyReliableTransferHid:" + std::to_string(pr_hid),
      "",
  };

  return join_lines(lines, "\n");
}

}  // namespace opennow
